gustaf/io/meshio.py

```diff
@@ -18,58 +19,96 @@
     """
-    mesh_type = Faces
-
     fname = pathlib.Path(fname)
     if not (fname.exists() and fname.is_file()):
         raise ValueError(
             "The given file does not point to file. The given path is: "
             f"{fname.resolve()}"
         )
 
+    # Read mesh
     meshio_mesh: meshio.Mesh = meshio.read(fname)
     vertices = meshio_mesh.points
 
-    # check for 2D mesh
-    # Try for triangle grid
-    cells = meshio_mesh.get_cells_type("triangle")
-    # If  no triangle elements, try for square
-    if len(cells) == 0:
-        cells = meshio_mesh.get_cells_type("quad")
-    if not len(cells) > 0:
-        # 3D mesh
-        mesh_type = Volumes
-        cells = meshio_mesh.get_cells_type("tetra")
-        if len(cells) == 0:
-            cells = meshio_mesh.get_cells_type("hexahedron")
-
-    for i, cell in enumerate(cells):
-        if i == 0:
-            elements = cell.data
-        else:
-            elements = np.vstack((elements, cell.data))
-
-    mesh = mesh_type(vertices=vertices, elements=elements)
+    # Maps meshio element types to gustaf types and dimensionality
+    # Meshio vertex elements are ommited, as they do not follow gustaf logics
+    meshio_mapping = {
+        "hexahedron": [Volumes, 3],
+        "tetra": [Volumes, 3],
+        "quad": [Faces, 2],
+        "triangle": [Faces, 2],
+        "line": [Edges, 1],
+    }
+    meshes = [
+        [
+            meshio_mapping[key][0](
+                vertices=vertices, elements=meshio_mesh.cells_dict[key]
+            ),
+            meshio_mapping[key][1],
+        ]
+        for key in meshio_mapping.keys()
+        if key in meshio_mesh.cells_dict.keys()
+    ]
+    # Sort by decreasing dimensionality
+    meshes.sort(key=lambda x: -x[1])
+
+    # Raises exception if the mesh file contains gustaf-incompartible types
+    for elem_type in meshio_mesh.cells_dict.keys():
+        if elem_type not in list(meshio_mapping.keys()) + ["vertex"]:
+            # Be aware that gustaf currently only supports linear element
+            # types.
+            raise NotImplementedError(
+                f"Sorry, {elem_type} elements currently are not supported."
+            )
 
-    return mesh
+    if set_boundary is True:
+        # mesh.BC is only defined for volumes and faces
+        for i in range(len(meshes)):
+            # Indicator for boundary
+            if meshes[i][1] in [3, 2]:
+                for j in range(i, len(meshes)):
+                    # Considers only (d-1)-dimensional subspaces
+                    if meshes[j][1] == meshes[i][1] - 1:
+                        meshes[i][0].BC[
+                            f"{meshes[j][0].whatami}-nodes"
+                        ] = np.unique(meshes[j][0].elements)
+
+    # Ensures backwards-compartibility
+    if return_only_one_mesh:
+        # Return highest-dimensional mesh
+        return_value = meshes[0][0]
+    else:
+        return_value = [mesh[0] for mesh in meshes]
```

Comment:
```suggestion
    return_value = [mesh[0] for mesh in meshes]
```
You can just return the whole thing without `return_only_one_mesh` option